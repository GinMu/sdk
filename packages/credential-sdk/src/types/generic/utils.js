import TypedArray from './typed-array';
import TypedBytesArray from './typed-bytes-array';

export class TypedArrayOfBytesArrays extends TypedArray {
  static Class = TypedBytesArray;
}

/**
 * Creates a converter function that processes an item through multiple parsers.
 *
 * @param {Function} finalize - Function to apply to each processed item
 * @param {Class} parserClass - Primary parser class used to parse the initial item
 * @param {...Function} parsers - Additional parser functions to apply
 * @returns {Function} A function that takes an item and returns processed results
 */
export const createConverter = (finalize, parserClass, ...parsers) => (item) => {
  // Initialize result array with original item
  const res = [item];

  try {
    // First parse the item using primary parser
    const parsedDid = parserClass.from(item);

    // Then pass the parsed result through additional parsers
    for (const parser of parsers) {
      try {
        res.push(parser.from(parsedDid));
      } catch {
        // Error is intentionally ignored
      }
    }
  } catch {
    // Initial parsing error is intentionally ignored
  }

  // Apply finalization to all processed results
  return res.map(finalize);
};
