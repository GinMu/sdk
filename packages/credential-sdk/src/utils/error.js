/**
 * Throws an error with the provided message or generated by a function.
 * @param {function | string | Error} err - The error message, function returning an error, or an Error object.
 */
export const panic = (error) => {
  if (typeof error === 'function') {
    return panic(error());
  }

  throw error instanceof Error ? error : new Error(error);
};

/**
 * Wraps a function to prefix its error messages with a specified string.
 * If an error is thrown by the wrapped function, this error will be caught,
 * modified to include the provided prefix in its message, and then rethrown.
 *
 * @param {function(): T} fn - The function to wrap
 * @param {string} prefix - Prefix to add before the original error message
 * @returns {T} The return value of the wrapped function
 *
 * @example
 * const wrappedFn = catchErrorWith('Error in component', () => {
 *   throw new Error('Something went wrong');
 * });
 *
 * try {
 *   wrappedFn();
 * } catch (err) {
 *   console.error(err.message); // "Error in component: Something went wrong"
 * }
 */
export const catchErrorWith = (prefix, fn) => {
  try {
    return fn();
  } catch (err) {
    err.message = `${prefix}: ${err.message}`;

    throw err;
  }
};

/**
 * Creates a function that will prefix any error messages from the original
 * function call with the specified string. This is useful for adding context
 * to errors thrown by functions that don't directly handle their own error messaging.
 *
 * @param {function} fn - The function to wrap
 * @param {string} prefix - Prefix to add before any error messages
 * @returns {function(...args): T} A new function that will execute the original function with provided arguments and prefix its errors
 */
export const catchFnErrorWith = (prefix, fn) => (...args) => catchErrorWith(prefix, () => fn(...args));
