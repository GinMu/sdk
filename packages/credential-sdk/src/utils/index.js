export * from './types';
export * from './interfaces';
export * from './async';
export * from './json-fetch';
export * from './inheritance';
export * from './encoding';
export * from './pattern-matcher';
export * from './ident';
export * from './error';
export * from './assert';
