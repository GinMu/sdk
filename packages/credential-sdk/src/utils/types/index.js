export * from './array';
export * from './number';
export * from './promise';
export * from './iterable';
export * from './map';
export * from './object';
export * from './string';
export * from './set';
export * from './bigint';
export * from './bytes';
