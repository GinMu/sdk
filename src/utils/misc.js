import elliptic from 'elliptic';
import { blake2AsHex, randomAsHex } from '@polkadot/util-crypto';

import { sha256 } from 'js-sha256';
import {
  PublicKeyEd25519,
  PublicKeySecp256k1,
  PublicKeySr25519, // eslint-disable-line
} from '../public-keys';
import {
  SignatureEd25519,
  SignatureSecp256k1,
  SignatureSr25519, // eslint-disable-line
} from '../signatures';

const EC = elliptic.ec;
const secp256k1Curve = new EC('secp256k1');

/** // TODO: Error handling when `stateChange` is not registered
 * Helper function to return bytes of a `StateChange` enum. Updates like key change, DID removal, revocation, etc
 * require the change to be wrapped in `StateChange` before serializing for signing.
 * @param {object} api - Promise API from polkadot-js
 * @param {object} stateChange - A representation of a `StateChange` enum variant
 * @return {array} An array of Uint8
 */
export function getBytesForStateChange(api, stateChange) {
  return api.createType('StateChange', stateChange).toU8a();
}

export function getStateChange(api, name, value) {
  const stateChange = {};
  stateChange[name] = value;
  return getBytesForStateChange(api, stateChange);
}

/**
 * Generate keypair for Ecdsa over Secp256k1. Explicitly denying other options to keep the API simple
 * @param {string} [pers] - A string
 * @param {array|string} [entropy] - A byte array or hex string
 * @returns {object} A keypair
 */

export function generateEcdsaSecp256k1Keypair(entropy = null) {
  return secp256k1Curve.genKeyPair({ entropy });
}

/**
 * Verify a given signature on a given message
 * @param {array} message - Bytes of message. Its assumed that the message is not hashed already
 * and hashed before verifying
 * @param {SignatureSecp256k1} signature - signature to verify
 * @param {PublicKeySecp256k1} publicKey - Secp256k1 public key for verification
 * @returns {boolean} True when signature is valid, false otherwise
 */
export function verifyEcdsaSecp256k1Sig(message, signature, publicKey) {
  const hash = sha256.digest(message);
  return verifyEcdsaSecp256k1SigPrehashed(hash, signature, publicKey);
}

/**
 * Verify a given signature on a given message hash
 * @param {array} messageHash - Hash of the message. Its assumed that the message is hashed already
 * @param {SignatureSecp256k1} signature - signature to verify
 * @param {PublicKeySecp256k1} publicKey - Secp256k1 public key for verification
 * @returns {boolean} True when signature is valid, false otherwise
 */
export function verifyEcdsaSecp256k1SigPrehashed(
  messageHash,
  signature,
  publicKey,
) {
  // Remove the leading `0x`
  const sigHex = signature.value.slice(2);
  // Break it in 2 chunks of 32 bytes each
  const sig = { r: sigHex.slice(0, 64), s: sigHex.slice(64, 128) };
  // Remove the leading `0x`
  const pkHex = publicKey.value.slice(2);
  // Generate public key object. Not extracting the public key for signature as the verifier
  // should always know what public key is being used.
  const pk = secp256k1Curve.keyFromPublic(pkHex, 'hex');
  return secp256k1Curve.verify(messageHash, sig, pk);
}

/**
 * Return the type of signature from a given keypair
 * @param {object} pair - Can be a keypair from polkadot-js or elliptic library.
 * @returns {string|*}
 */
export function getKeyPairType(pair) {
  if (pair.type) {
    return pair.type;
  }

  if (pair.ec && pair.priv) {
    // elliptic library's pair has `ec`, `priv` and `pub`. There is not a cleaner way to detect that
    return 'secp256k1';
  }
  throw new Error('Cannot detect key pair type');
}

/**
 * Inspect the `type` of the `KeyringPair` to generate the correct kind of PublicKey.
 * @param {object} pair - A polkadot-js KeyringPair.
 * @return {PublicKey} An instance of the correct subclass of PublicKey
 */
export function getPublicKeyFromKeyringPair(pair) {
  const type = getKeyPairType(pair);
  let Cls;
  if (type === 'ed25519') {
    Cls = PublicKeyEd25519;
  } else if (type === 'sr25519') {
    Cls = PublicKeySr25519;
  } else {
    Cls = PublicKeySecp256k1;
  }
  return Cls.fromKeyringPair(pair);
}

/**
 * Inspect the `type` of the `KeyringPair` to generate the correct kind of Signature.
 * @param {object} pair - A polkadot-js KeyringPair.
 * @param {array} message - an array of bytes (Uint8)
 * @returns {Signature} An instance of the correct subclass of Signature
 */
export function getSignatureFromKeyringPair(pair, message) {
  const type = getKeyPairType(pair);
  let Cls;
  if (type === 'ed25519') {
    Cls = SignatureEd25519;
  } else if (type === 'sr25519') {
    Cls = SignatureSr25519;
  } else {
    Cls = SignatureSecp256k1;
  }
  return new Cls(message, pair);
}

/**
 * Wrapped keypair used to interact with the dock blockchain.
 */
export class DockKeypair {
  /**
   * Wraps supplied keypair into a `DockKeypair`.
   *
   * @param {*} keyPair
   */
  constructor(keyPair) {
    this.keyPair = keyPair;
  }

  /**
   * Returns underlying public key.
   */
  publicKey() {
    return getPublicKeyFromKeyringPair(this.keyPair);
  }

  /**
   * Signs supplied message using underlying keypair.
   * @param {*} message
   */
  sign(message) {
    return getSignatureFromKeyringPair(this.keyPair, message);
  }

  /**
   * Generates random `Secp256k1` keypair.
   *
   * @param params
   * @returns {this}
   */
  static randomSecp256k1() {
    return new this(generateEcdsaSecp256k1Keypair(randomAsHex(32)));
  }
}

/**
 * Get unique elements from an array as seen by the filterCallback function.
 * @param {array} a - Array to check for duplicates.
 * @param {function} filterCallback - Elements will be fed to this function before comparison.
 * @returns {*}
 */
export function getUniqueElementsFromArray(a, filterCallback) {
  const seen = new Set();
  return a.filter((item) => {
    const k = filterCallback(item);
    return seen.has(k) ? false : seen.add(k);
  });
}

/**
 * Encodes an extrinsic as a blake2 hash
 * @param {*} api - API for creating Call type
 * @param {*} tx - Extrinsic to encode
 * @returns {string}
 */
export function encodeExtrinsicAsHash(api, tx) {
  return blake2AsHex(api.createType('Call', tx).toU8a());
}

/**
 * Get the nonce to be used for sending the next transaction if not provided already.
 * @param {DockDidOrDidMethodKey} didOrDidMethodKey - DID whose nonce is needed
 * @param nonce - If provided, returned as it is.
 * @param didModule - Reference to the DID module. If nonce is not provided then the next nonce for the DID is fetched by
 * using this
 * @returns {Promise<undefined|*>}
 */
export async function getDidNonce(
  didOrDidMethodKey,
  nonce = undefined,
  didModule = undefined,
) {
  if (nonce === undefined && didModule === undefined) {
    throw new Error(
      'Provide either nonce or didModule to fetch nonce but none provided',
    );
  }
  if (nonce === undefined) {
    return didModule.getNextNonceForDid(didOrDidMethodKey);
  }
  return nonce;
}

/**
 * Returns string containing comma separated items of the provided iterable.
 *
 * @template V
 * @param {Iterable<V>} iter
 * @returns {string}
 */
export const fmtIter = (iter) => `\`[${[...iter].map((item) => item.toString()).join(', ')}]\``;

const PatternAttrs = new Set([
  '$matchType',
  '$matchValue',
  '$matchObject',
  '$matchIterable',
  '$instanceOf',
  '$iterableOf',
  '$mapOf',
  '$anyOf',
  '$objOf',
]);

/**
 * Ensures that provided value matches supplied pattern, throws an error otherwise.
 *
 * @param pattern
 * @param value
 * @param path
 */
// eslint-disable-next-line sonarjs/cognitive-complexity
export const ensureMatchesStructurePattern = (pattern, value, path = []) => {
  const patternAttrs = new Set(Object.keys(pattern));

  for (const key of patternAttrs) {
    if (!PatternAttrs.has(key)) {
      throw new Error(
        `Invalid pattern modifier: ${key}, path: \`${path.join('.')}\`.`,
      );
    }
  }

  // eslint-disable-next-line valid-typeof
  if (patternAttrs.has('$matchType') && typeof value !== pattern.$matchType) {
    throw new Error(
      `Invalid value provided, expected value with type \`${
        pattern.$matchType
      }\`, received value with type \`${typeof value}\`, path: \`${path.join(
        '.',
      )}\`.`,
    );
  }

  if (patternAttrs.has('$matchValue') && value !== pattern.$matchValue) {
    throw new Error(
      `Unknown value \`${value}\`, expected ${
        pattern.$matchValue
      }, path: \`${path.join('.')}\`.`,
    );
  }

  if (patternAttrs.has('$matchObject')) {
    for (const key of Object.keys(value)) {
      if (!Object.hasOwnProperty.call(pattern.$matchObject, key)) {
        throw new Error(
          `Invalid property \`${key}\`, expected keys: \`${fmtIter(
            Object.keys(pattern.$matchObject),
          )}\`, path: \`${path.join('.')}\`, pattern: \`${JSON.stringify(
            pattern,
          )}\``,
        );
      }

      ensureMatchesStructurePattern(
        pattern.$matchObject[key],
        value[key],
        path.concat(key),
      );
    }
  }

  if (patternAttrs.has('$matchIterable')) {
    if (typeof value[Symbol.iterator] !== 'function') {
      throw new Error(
        `Iterable expected, received: ${value}, path: \`${path.join('.')}\`.`,
      );
    }
    const objectIter = value[Symbol.iterator]();

    let idx = 0;
    for (const pat of pattern.$matchIterable) {
      const { value: item, done } = objectIter.next();
      if (done) {
        throw new Error(
          `Value iterable is shorter than expected, received: ${fmtIter(
            value,
          )}, path: \`${path.join('.')}\`.`,
        );
      }

      ensureMatchesStructurePattern(pat, item, path.concat(`@item#${idx++}`));
    }
  }

  if (
    patternAttrs.has('$instanceOf')
    && !(value instanceof pattern.$instanceOf)
  ) {
    throw new Error(
      `Invalid value provided, expected instance of \`${
        pattern.$instanceOf.name
      }\`, received instance of \`${
        value?.constructor?.name
      }\`, path: \`${path.join('.')}\`.`,
    );
  }

  if (patternAttrs.has('$iterableOf')) {
    if (typeof value?.[Symbol.iterator] !== 'function') {
      throw new Error(
        `Iterable expected, received \`${value}\`, path: \`${path.join('.')}\`.`,
      );
    }

    let idx = 0;
    for (const entry of value) {
      ensureMatchesStructurePattern(
        pattern.$iterableOf,
        entry,
        path.concat(`@item#${idx++}`),
      );
    }
  }

  if (patternAttrs.has('$mapOf')) {
    if (typeof value?.entries !== 'function') {
      throw new Error(
        `Map expected, received \`${value}\`, path: \`${path.join('.')}\`.`,
      );
    }

    if (!Array.isArray(pattern.$mapOf) || pattern.$mapOf.length !== 2) {
      throw new Error(
        `\`$mapOf\` pattern should be an array with two items, received \`${JSON.stringify(
          pattern,
        )}\`, path: \`${path.join('.')}\``,
      );
    }

    for (const [key, item] of value.entries()) {
      ensureMatchesStructurePattern(
        pattern.$mapOf[0],
        key,
        path.concat(`${key}#key`),
      );
      ensureMatchesStructurePattern(pattern.$mapOf[1], item, path.concat(key));
    }
  }

  if (patternAttrs.has('$anyOf')) {
    let anySucceeded = false;
    const errors = [];
    for (const patEntry of pattern.$anyOf) {
      try {
        ensureMatchesStructurePattern(patEntry, value, path);
        anySucceeded = true;
      } catch (err) {
        errors.push(err);
      }
    }

    if (!anySucceeded) {
      throw new Error(
        `Neither of pattern succeeded for \`${value}\`: ${JSON.stringify(
          errors.map((err) => err.message),
        )}, path: \`${path.join('.')}\`.`,
      );
    }
  }

  if (patternAttrs.has('$objOf')) {
    const keys = Object.keys(value);
    if (keys.length !== 1) {
      throw new Error('Expected a single key');
    }
    if (!Object.hasOwnProperty.call(pattern.$objOf, keys[0])) {
      throw new Error(
        `Invalid value key provided, expected one of \`${fmtIter(
          Object.keys(pattern.$objOf),
        )}\`, received \`${keys[0]}\`, path: \`${path.join('.')}\`.`,
      );
    }
    ensureMatchesStructurePattern(
      pattern.$objOf[keys[0]],
      value[keys[0]],
      path.concat(keys[0]),
    );
  }
};
