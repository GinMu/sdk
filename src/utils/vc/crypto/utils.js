/**
 * Creates a function to convert a derived proof credential to the native presentation
 * format with the provided type.
 * @param {*} proofType
 * @returns {function(object): object}
 */
export const buildPresentationConverter = (proofType) => function convertToPresentation(document) {
  const {
    '@context': context,
    type,
    credentialSchema,
    issuer,
    issuanceDate,
    proof,
    ...revealedAttributes
  } = document;

  return {
    version: '0.1.0',
    nonce: proof.nonce,
    context: proof.context,
    spec: {
      credentials: [
        {
          version: proof.version,
          schema: JSON.stringify(credentialSchema),
          revealedAttributes: {
            proof: {
              type: proofType,
              verificationMethod: proof.verificationMethod,
            },
            '@context': JSON.stringify(context),
            type: JSON.stringify(type),
            ...revealedAttributes,
          },
        },
      ],
      attributeEqualities: proof.attributeEqualities,
    },
    attributeCiphertexts: proof.attributeCiphertexts,
    proof: proof.proofValue,
  };
};
