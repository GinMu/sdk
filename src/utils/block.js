import { BlockHash, BlockNumber } from "@polkadot/types/interfaces"; // eslint-disable-line
import { SignedBlockExtended } from "@polkadot/types/types"; // eslint-disable-line
import { ApiPromise } from "@polkadot/api"; // eslint-disable-line
import { ReusablePromiseMap, ReusablePromise } from './async';

/**
 * Fetches and caches blocks by their hashes and optionally numbers.
 * `finalized` flag determines the last hash/number querying algorithms
 * and whether blocks will be cached by their numbers or not.
 */
export class BlocksProvider {
  /**
   *
   * @param {object} configuration
   * @param {ApiPromise} [configuration.api]
   * @param {boolean} [configuration.cacheCapacity=100]
   * @param {boolean} [configuration.finalized=false]
   */
  constructor({ api, cacheCapacity = 100, finalized = false } = {}) {
    this.api = api;
    this.finalized = finalized;

    this.lastHashCall = new ReusablePromise();
    this.blockByHashCalls = new ReusablePromiseMap({
      capacity: cacheCapacity,
      save: true,
    });
    this.blockByNumberCalls = new ReusablePromiseMap({
      capacity: cacheCapacity,
      save: finalized,
    });

    this.maxBlockNumber = 0;
  }

  /**
   * Returns the latest available hash for the block (finalized if `finalized=true`).
   *
   * @returns {Promise<BlockHash>}
   */
  async lastHash() {
    return await this.lastHashCall.call(() => (this.finalized
      ? this.api.rpc.chain.getFinalizedHead()
      : this.api.rpc.chain.getBlockHash()));
  }

  /**
   * Returns the latest available number for the block (finalized if `finalized=true`).
   *
   * @returns {Promise<BlockNumber>}
   */
  async lastNumber() {
    const hash = await this.lastHash();

    const {
      block: {
        header: { number },
      },
    } = await this.blockByHash(hash, { skipCheck: true });
    this.maxBlockNumber = Math.max(number.toNumber(), this.maxBlockNumber);

    return number;
  }

  /**
   * Retrieves blocks by hash from either the local cache or by querying the node.
   *
   * @param {BlockHash|string} hash
   * @param {object} configuration
   * @param {boolean} [configuration.skipCheck=false]
   * @returns {Promise<SignedBlockExtended>}
   */
  async blockByHash(hash, { skipCheck = false } = {}) {
    const block = await this.blockByHashCalls.callByKey(String(hash), () => this.api.derive.chain.getBlock(hash));
    const {
      block: {
        header: { number },
      },
    } = block;

    if (!skipCheck && number.toNumber() > this.maxBlockNumber) {
      await this.lastNumber();

      if (number.toNumber() > this.maxBlockNumber) {
        throw new Error(
          `Provided block's number ${number.toNumber()} is higher than the latest known block number ${this.maxBlockNumber}`,
        );
      }
    }

    return block;
  }

  /**
   * Retrieves blocks by number from either the local cache or by querying the node.
   *
   * @param {BlockNumber|number} number
   * @returns {Promise<SignedBlockExtended>}
   */
  async blockByNumber(number) {
    return await this.blockByNumberCalls.callByKey(String(number), async () => {
      const hash = await this.api.rpc.chain.getBlockHash(number);
      return await this.blockByHash(hash);
    });
  }
}
