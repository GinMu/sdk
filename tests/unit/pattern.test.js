import { ensureMatchesStructurePattern } from '../../src/utils/misc';

describe('ensureMatchesStructurePattern', () => {
  test('$matchType', () => {
    const pat = { $matchType: 'number' };
    expect(() => ensureMatchesStructurePattern(pat, 'sasda')).toThrowErrorMatchingSnapshot();
    ensureMatchesStructurePattern(pat, 10);
  });

  test('$matchValue', () => {
    const pat = { $matchValue: 11 };

    expect(() => ensureMatchesStructurePattern(pat, 10)).toThrowErrorMatchingSnapshot();
    ensureMatchesStructurePattern(pat, 11);
  });

  test('$matchObject', () => {
    const pat = {
      $matchObject: { a: { $matchValue: 1 }, b: { $matchValue: 2 } },
    };

    expect(() => ensureMatchesStructurePattern(pat, { c: 3 })).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, { a: 3 })).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, { a: 1, b: 3 })).toThrowErrorMatchingSnapshot();

    ensureMatchesStructurePattern(pat, { a: 1 });
    ensureMatchesStructurePattern(pat, { b: 2 });
    ensureMatchesStructurePattern(pat, { a: 1, b: 2 });
  });

  test('$matchIterable', () => {
    const pat = { $matchIterable: [{ $matchValue: 1 }, { $matchValue: 2 }] };
    expect(() => ensureMatchesStructurePattern(pat, [3])).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, [1, 3])).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, [])).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, [1])).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, [2])).toThrowErrorMatchingSnapshot();

    expect(ensureMatchesStructurePattern(pat, [1, 2]));
  });

  test('$instanceOf', () => {
    expect(() => ensureMatchesStructurePattern({ $instanceOf: Function }, {})).toThrowErrorMatchingSnapshot();
    ensureMatchesStructurePattern({ $instanceOf: Object }, {});
  });

  test('$iterableOf', () => {
    const pat = { $iterableOf: { $matchValue: 1 } };

    expect(() => ensureMatchesStructurePattern(pat, [3])).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, [1, 3])).toThrowErrorMatchingSnapshot();

    expect(() => ensureMatchesStructurePattern(pat, [2])).toThrowErrorMatchingSnapshot();

    expect(() => ensureMatchesStructurePattern(pat, [1, 2])).toThrowErrorMatchingSnapshot();

    ensureMatchesStructurePattern(pat, []);
    ensureMatchesStructurePattern(pat, [1]);
    ensureMatchesStructurePattern(pat, [1, 1, 1, 1, 1]);
  });

  test('$mapOf', () => {
    const pat = { $mapOf: [{ $matchValue: 1 }, { $matchValue: 2 }] };

    expect(() => ensureMatchesStructurePattern(pat, new Map([[1, 3]]))).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, new Map([[3, 2]]))).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(
      pat,
      new Map([
        [1, 2],
        [3, 1],
      ]),
    )).toThrowErrorMatchingSnapshot();

    ensureMatchesStructurePattern(pat, new Map([]));
    ensureMatchesStructurePattern(pat, new Map([[1, 2]]));
  });

  test('$anyOf', () => {
    const pat = { $anyOf: [{ $matchValue: 1 }, { $matchValue: 2 }] };

    expect(() => ensureMatchesStructurePattern(pat, 3)).toThrowErrorMatchingSnapshot();
    ensureMatchesStructurePattern(pat, 1);
    ensureMatchesStructurePattern(pat, 2);
  });

  test('$mapOf', () => {
    const pat = { $objOf: { A: { $matchValue: 1 } } };

    expect(() => ensureMatchesStructurePattern(pat, { A: 5 })).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, { B: 1 })).toThrowErrorMatchingSnapshot();
    expect(() => ensureMatchesStructurePattern(pat, { A: 1, B: 3 })).toThrowErrorMatchingSnapshot();

    ensureMatchesStructurePattern(pat, { A: 1 });
  });
});
