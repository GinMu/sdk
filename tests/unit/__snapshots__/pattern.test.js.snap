// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensureMatchesPattern $anyOf 1`] = `"Neither of patterns succeeded for \`3\`"`;

exports[`ensureMatchesPattern $anyOf 2`] = `Array []`;

exports[`ensureMatchesPattern $anyOf 3`] = `
Object {
  "$anyOf": Array [
    Object {
      "$matchValue": 1,
    },
    Object {
      "$matchValue": 2,
    },
  ],
}
`;

exports[`ensureMatchesPattern $anyOf 4`] = `
Array [
  [Error: Unknown value \`3\`, expected 1],
  [Error: Unknown value \`3\`, expected 2],
]
`;

exports[`ensureMatchesPattern $instanceOf 1`] = `"Invalid value provided, expected instance of \`Function\`, received instance of \`Object\`"`;

exports[`ensureMatchesPattern $instanceOf 2`] = `Array []`;

exports[`ensureMatchesPattern $instanceOf 3`] = `
Object {
  "$instanceOf": [Function],
}
`;

exports[`ensureMatchesPattern $instanceOf 4`] = `Array []`;

exports[`ensureMatchesPattern $iterableOf 1`] = `"Unknown value \`3\`, expected 1, path: \`0\`"`;

exports[`ensureMatchesPattern $iterableOf 2`] = `
Array [
  0,
]
`;

exports[`ensureMatchesPattern $iterableOf 3`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $iterableOf 4`] = `Array []`;

exports[`ensureMatchesPattern $iterableOf 5`] = `"Unknown value \`3\`, expected 1, path: \`1\`"`;

exports[`ensureMatchesPattern $iterableOf 6`] = `
Array [
  1,
]
`;

exports[`ensureMatchesPattern $iterableOf 7`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $iterableOf 8`] = `Array []`;

exports[`ensureMatchesPattern $iterableOf 9`] = `"Unknown value \`2\`, expected 1, path: \`0\`"`;

exports[`ensureMatchesPattern $iterableOf 10`] = `
Array [
  0,
]
`;

exports[`ensureMatchesPattern $iterableOf 11`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $iterableOf 12`] = `Array []`;

exports[`ensureMatchesPattern $iterableOf 13`] = `"Unknown value \`2\`, expected 1, path: \`1\`"`;

exports[`ensureMatchesPattern $iterableOf 14`] = `
Array [
  1,
]
`;

exports[`ensureMatchesPattern $iterableOf 15`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $iterableOf 16`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 1`] = `"Unknown value \`3\`, expected 2, path: \`1\`"`;

exports[`ensureMatchesPattern $mapOf 2`] = `
Array [
  1,
]
`;

exports[`ensureMatchesPattern $mapOf 3`] = `
Object {
  "$matchValue": 2,
}
`;

exports[`ensureMatchesPattern $mapOf 4`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 5`] = `"Unknown value \`3\`, expected 1, path: \`3#key\`"`;

exports[`ensureMatchesPattern $mapOf 6`] = `
Array [
  "3#key",
]
`;

exports[`ensureMatchesPattern $mapOf 7`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $mapOf 8`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 9`] = `"Unknown value \`3\`, expected 1, path: \`3#key\`"`;

exports[`ensureMatchesPattern $mapOf 10`] = `
Array [
  "3#key",
]
`;

exports[`ensureMatchesPattern $mapOf 11`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $mapOf 12`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 13`] = `"Unknown value \`5\`, expected 1, path: \`A\`"`;

exports[`ensureMatchesPattern $mapOf 14`] = `
Array [
  "A",
]
`;

exports[`ensureMatchesPattern $mapOf 15`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $mapOf 16`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 17`] = `"Invalid value key provided, expected one of \`\`[A]\`\`, received \`B\`"`;

exports[`ensureMatchesPattern $mapOf 18`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 19`] = `
Object {
  "$objOf": Object {
    "A": Object {
      "$matchValue": 1,
    },
  },
}
`;

exports[`ensureMatchesPattern $mapOf 20`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 21`] = `"Expected a single key"`;

exports[`ensureMatchesPattern $mapOf 22`] = `Array []`;

exports[`ensureMatchesPattern $mapOf 23`] = `
Object {
  "$objOf": Object {
    "A": Object {
      "$matchValue": 1,
    },
  },
}
`;

exports[`ensureMatchesPattern $mapOf 24`] = `Array []`;

exports[`ensureMatchesPattern $matchIterable 1`] = `"Unknown value \`3\`, expected 1, path: \`0\`"`;

exports[`ensureMatchesPattern $matchIterable 2`] = `
Array [
  0,
]
`;

exports[`ensureMatchesPattern $matchIterable 3`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $matchIterable 4`] = `Array []`;

exports[`ensureMatchesPattern $matchIterable 5`] = `"Unknown value \`3\`, expected 2, path: \`1\`"`;

exports[`ensureMatchesPattern $matchIterable 6`] = `
Array [
  1,
]
`;

exports[`ensureMatchesPattern $matchIterable 7`] = `
Object {
  "$matchValue": 2,
}
`;

exports[`ensureMatchesPattern $matchIterable 8`] = `Array []`;

exports[`ensureMatchesPattern $matchIterable 9`] = `"Value iterable is shorter than expected, received: \`[]\`"`;

exports[`ensureMatchesPattern $matchIterable 10`] = `Array []`;

exports[`ensureMatchesPattern $matchIterable 11`] = `
Object {
  "$matchIterable": Array [
    Object {
      "$matchValue": 1,
    },
    Object {
      "$matchValue": 2,
    },
  ],
}
`;

exports[`ensureMatchesPattern $matchIterable 12`] = `Array []`;

exports[`ensureMatchesPattern $matchIterable 13`] = `"Value iterable is shorter than expected, received: \`[1]\`"`;

exports[`ensureMatchesPattern $matchIterable 14`] = `Array []`;

exports[`ensureMatchesPattern $matchIterable 15`] = `
Object {
  "$matchIterable": Array [
    Object {
      "$matchValue": 1,
    },
    Object {
      "$matchValue": 2,
    },
  ],
}
`;

exports[`ensureMatchesPattern $matchIterable 16`] = `Array []`;

exports[`ensureMatchesPattern $matchIterable 17`] = `"Unknown value \`2\`, expected 1, path: \`0\`"`;

exports[`ensureMatchesPattern $matchIterable 18`] = `
Array [
  0,
]
`;

exports[`ensureMatchesPattern $matchIterable 19`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $matchIterable 20`] = `Array []`;

exports[`ensureMatchesPattern $matchObject 1`] = `"Invalid property \`c\`, expected keys: \`\`[a, b]\`\`"`;

exports[`ensureMatchesPattern $matchObject 2`] = `Array []`;

exports[`ensureMatchesPattern $matchObject 3`] = `
Object {
  "$matchObject": Object {
    "a": Object {
      "$matchValue": 1,
    },
    "b": Object {
      "$matchValue": 2,
    },
  },
}
`;

exports[`ensureMatchesPattern $matchObject 4`] = `Array []`;

exports[`ensureMatchesPattern $matchObject 5`] = `"Unknown value \`3\`, expected 1, path: \`a\`"`;

exports[`ensureMatchesPattern $matchObject 6`] = `
Array [
  "a",
]
`;

exports[`ensureMatchesPattern $matchObject 7`] = `
Object {
  "$matchValue": 1,
}
`;

exports[`ensureMatchesPattern $matchObject 8`] = `Array []`;

exports[`ensureMatchesPattern $matchObject 9`] = `"Unknown value \`3\`, expected 2, path: \`b\`"`;

exports[`ensureMatchesPattern $matchObject 10`] = `
Array [
  "b",
]
`;

exports[`ensureMatchesPattern $matchObject 11`] = `
Object {
  "$matchValue": 2,
}
`;

exports[`ensureMatchesPattern $matchObject 12`] = `Array []`;

exports[`ensureMatchesPattern $matchType 1`] = `"Invalid value provided, expected value with type \`number\`, received value with type \`string\`"`;

exports[`ensureMatchesPattern $matchType 2`] = `Array []`;

exports[`ensureMatchesPattern $matchType 3`] = `
Object {
  "$matchType": "number",
}
`;

exports[`ensureMatchesPattern $matchType 4`] = `Array []`;

exports[`ensureMatchesPattern $matchValue 1`] = `"Unknown value \`10\`, expected 11"`;

exports[`ensureMatchesPattern $matchValue 2`] = `Array []`;

exports[`ensureMatchesPattern $matchValue 3`] = `
Object {
  "$matchValue": 11,
}
`;

exports[`ensureMatchesPattern $matchValue 4`] = `Array []`;
